windows/src/global/delphi/general/Keyman.System.AndroidStringToKeymanLocaleString.pas

```diff
@@ -0,0 +1,67 @@
+unit Keyman.System.AndroidStringToKeymanLocaleString;
+
+interface
+
+type
+  TAndroidStringToKeymanLocaleString = class sealed
+  public
+    class function Transform(text: string): string; static;
+  end;
+
+implementation
+
+uses
+  System.SysUtils;
+
+class function TAndroidStringToKeymanLocaleString.Transform(text: string): string;
+var
+  i, j: Integer;
+begin
+  // strings.xml is not completely identical to the older locale.xml format we use
+
+  // 1. Trim the text first (to avoid this, use "  text  ")
+  text := text.Trim;
+
+  // 2. remove surrounding quotes from string
+  if text.StartsWith('"') and text.EndsWith('"') then
+  begin
+    text := text.Substring(1, text.Length-2);
+  end;
+
+  // 3. Remove escapes
+  i := 1; j := 1;
+  SetLength(Result, text.Length);
+  while i <= text.Length do
+  begin
+    if text[i] = '\' then
+    begin
+      Inc(i);
+      case text[i] of
+        't': Result[j] := #9;
+        'n': begin Result[j] := #13; Inc(j); Result[j] := #10; end;
+        else Result[j] := text[i];
+      end;
+    end
+    else if (i <= text.Length - 3) and
+      (text[i] = '%') and
+      (text[i+1] >= '1') and (text[i+1] <= '9') and
+      (text[i+2] = '$') and
+      CharInSet(text[i+3], ['d','u','e','f','g','n','m','p','s','x']) then
+    begin
+      // 4. Transform positional parameters
+      // convert from %1$d to %0:d; Delphi parameters are 0-based
+      // note: only supports up to 9 positional parameters
```

Comment:
Is the limit of 9 positional parameters made for practicality purposes?

Can maybe add a unit test for this boundary so we'll see a fail if someone tries to make a string with 10 params

Reply:
I don't think it's necessary. We use at most 2 parameters currently. It'd be overkill to create a test against all the strings.xml files anyway to exercise that!